### ISSUE TITLE
Qwen3.5 397B (AMD): async kernel-ordering race corrupts output + hangs (P0)

### ISSUE THREAD

**Issue by raikonenfnu** [raikonenfnu/tokenspeed#1]
## Summary

On **Qwen3.5-397B-A17B-MXFP4 (AMD/gfx950)**, running in the **normal asynchronous** execution config, the model **produces a wrong first token and then hangs** in decode. The exact same prompt is correct when launches are forced synchronous (`HIP_LAUNCH_BLOCKING=1`). A bug that disappears under synchronous launches is, by construction, a GPU **stream-ordering / synchronization** bug — a data dependency between two ops is not enforced with an event / `record_stream` / `synchronize`.

This is the **primary remaining blocker (P0)** for functional 397B serving. It reproduces at **TP=4 and TP=8**; it does **not** reproduce on 35B (TP=1). It is independent of (and not caused by) the committed int64 KV-cache overflow fix.

## Smallest reproducer

TP=4, default everything (CUDA graphs on, allreduce fusion on, async launches):

```bash
cd /home/stwinata/nod/sgl_runner/amd-Qwen3.5-397B-A17B-MXFP4
TP=4 PORT=30000 bash run_tokenspeed.sh > server_tp4.log 2>&1 &
PORT=$(grep -aoE 'gRPC server listening on 127.0.0.1:[0-9]+' server_tp4.log | tail -1 | grep -oE '[0-9]+$')
ENGINE_ADDR=127.0.0.1:$PORT python trigger_one.py   # "The capital of France is", greedy
```

- **Broken (async):** first token `4262` = `'PP'` (wrong), then hangs.
- **Correct (sync):** add `HIP_LAUNCH_BLOCKING=1 --enforce-eager` → first token `11751` = `' Paris'`, no hang (just slow).

py-spy shows every rank wedged at the start of the next step:
```
reset_states            (runtime_states.py:79)
reset_valid_cache_length (model_executor.py:1223)
_dispatch_forward        (engine/event_loop.py:630)
event_loop_overlap       (engine/event_loop.py:1489)
```

## Leading hypothesis

The overlap event loop runs the next forward on one stream while post-processing the previous step on another. `reset_states` resets/reuses per-step state buffers on stream A while a previous-step kernel still reads/writes them on stream B, with no event/`record_stream` enforcing "reset only after the consumer completes." Corruption vs hang are the two outcomes of the same race.

## Suggested next steps

1. **Localize:** binary-search a `torch.cuda.synchronize()` through the overlap loop / forward to find the first op whose async launch corrupts/wedges; or use selective `HIP_LAUNCH_BLOCKING`.
2. **Add the missing dependency:** `record_stream` and/or a CUDA event so the reset/reuse cannot begin before the prior step's consumers finish.
3. **Fix `reset_states` buffer lifetime:** double-buffer the state tensor or run the reset on the consumers' stream.
4. **Cheap triage:** disable the overlap loop — if that fixes it, the race is confirmed in the overlap stream handoff.
5. **Stopgap:** ship the synchronous config (`HIP_LAUNCH_BLOCKING=1 --enforce-eager` [+ `TOKENSPEED_DISABLE_ALLREDUCE_FUSION=1` at TP=8]) as the supported validation path (correct but slow).

## Code to audit

- `runtime/engine/event_loop.py` — stream/event mgmt around `_dispatch_forward` (l.630), `_commit_forward_results` (l.1071).
- `runtime/states/runtime_states.py:79` (`reset_states`), `runtime/.../model_executor.py:1223` (`reset_valid_cache_length`).
- Custom kernels first exercised on 397B prefill: GDN Triton chunk path, MoE MXFP4 (`ops/moe/triton/mxfp4.py`, `ops/moe/gluon`), `fused_mamba_state_copy`, fused qkv/rotary.

---
_Filed from local engineering notes (`~/agentic_results/distributed_qwen35_async_issue.md`, cross-ref `todo_qwen35_tokenspeed.md`). Status as of 2026-06-23._

**Comment by raikonenfnu** [comment]
## Root-caused & fixed ✅ (AMD MI350X / gfx950, branch `raikonen/gdn_rebased`)

The corruption is **not** in `reset_states` / the engine state buffers — that py-spy frame is just where the CPU wedges waiting on a GPU that's already stuck/corrupt. The real bug is a missing **trailing cross-rank barrier** in the Iris fused allreduce+RMSNorm kernel.

### How it was localized
Ran a clean fusion × overlap × eager matrix at TP=4 (prompt `"The capital of France is"`, greedy, first-token id):

| allreduce fusion | overlap | eager | first token | verdict |
|---|---|---|---|---|
| ON | ON  | off | `947 ' sp'` | BROKEN |
| ON | ON  | on  | `427 ' r'`  | BROKEN |
| ON | OFF | off | `11751 ' Paris'` | **CORRECT** |

Key takeaways:
- Bug lives in the **overlap scheduler stream handoff**, and **`--enforce-eager` alone does NOT fix it** — so it's *not* CUDA-graph replay. The original `HIP_LAUNCH_BLOCKING=1 --enforce-eager` "fix" worked only because `HIP_LAUNCH_BLOCKING` *also* serializes the two streams, masking the race.
- It manifests iff the Iris fused path runs (`triton_allreduce_residual_rmsnorm` → `IrisAllReduceResidualRMSNorm.fused`).
- ⚠️ Separately discovered: the `TOKENSPEED_DISABLE_ALLREDUCE_FUSION=1` workaround **itself corrupts TP=4** (the non-fused NCCL allreduce+RMSNorm path is numerically wrong there), so it must not be used as a correctness baseline.

### Root cause
`IrisAllReduceResidualRMSNorm.fused()` reuses one persistent symmetric-heap buffer `_input_buf`:
```python
in_view.copy_(input_tensor)   # write my shard
self._ctx.device_barrier()    # LEADING barrier: all ranks wrote before peers read
kernel[grid](in_view, ...)    # each rank reads PEERS' _input_buf via iris.load
# <-- no trailing barrier
```
The leading barrier guarantees write-before-read but nothing guarantees read-before-next-write: the next `fused()` call's `copy_` can clobber `_input_buf` while a **peer rank** is still `iris.load`-ing it. The non-overlap loop keeps ranks lock-stepped (CPU syncs each step on `copy_event`), hiding it; the overlap loop deliberately lets per-rank launch order diverge, opening the hazard window → cross-rank corrupted allreduce → wrong logits.

### Fix (`tokenspeed-kernel/.../ops/communication/iris.py`)
Add a trailing `self._ctx.device_barrier()` after the kernel so no rank overwrites `_input_buf` until all peers finish reading it. (+11 lines, no engine/event-loop changes.)

### Validation (default config: overlap ON + fusion ON + CUDA graphs, no workarounds)
- **TP=4** full coherence suite passes: `"The capital of France is Paris."`, primary colors, `def square(n): return n * n`, `17×4 = 68`, Rayleigh scattering, primes `2, 3, 5, 7, 11`. (Was total CJK/empty garbage before.)
- **TP=8** also coherent (also clears #2).

### Simplest repro / test
- Hardware: `TP=4 OVERLAP=1 EAGER=0 python repro_issue1.py | grep first_tok` → must be `11751 ' Paris'`.
- Focused multi-rank stress test: loop `IrisAllReduceResidualRMSNorm.fused()` with injected per-rank CPU skew and `allclose` vs reference allreduce+residual+RMSNorm — non-zero mismatches pre-fix, 0 post-fix.
